### PR TITLE
@(expr) in subprocess tokens

### LINKED ISSUE #2640: Python expression evaluation using @() not allowed as subexpression
I am trying to do something like this:

```
def upload_keys(user, server):
    scp /home/@(user)/.ssh/* @(user)@@(server):.ssh/.
```
Unfortunately it seems that the @() operator needs to be surrounded with spaces. 
```
def upload_keys(user, server):
    scp /home/ @(user) /.ssh/* @(user) @ @(server) :.ssh/.
```
This however doesn't work, because now the scp commands sees separate arguments.

A work around could be this
```
def upload_keys(user, server):
    $USER=user
    $SERVER=server
    scp /home/$USER/.ssh/* $USER@$SERVER:.ssh/.
```
But this pollutes the environment and results in a lot of copying.

Any clean solution to work with Python expressions as part of shell commands?

### PR DESCRIPTION
**Pull request by scopatz**
This closes #2640 

This was fun to implement.

### CODE REVIEW

File: xonsh/parsers/base.py [hunk 1]
```diff
@@ -3085,7 +3149,19 @@ def p_subproc_arg_single(self, p):
     def p_subproc_arg_many(self, p):
         """subproc_arg : subproc_arg subproc_arg_part"""
         # This glues the string together after parsing
-        p[0] = p[1] + p[2]
+        p1 = p[1]
+        p2 = p[2]
+        if isinstance(p1, ast.Str) and isinstance(p2, ast.Str):
+            p0 = ast.Str(p1.s + p2.s, lineno=p1.lineno, col_offset=p1.col_offset)
+        elif isinstance(p1, list):
```
Comment (gforsyth) [inline xonsh/parsers/base.py]: Is there a case here where both `p1` and `p2` are `list`s?  Does that need to be cased?

**Comment by melund** [comment]
This is just crazy :)   I can't wait to use the outer-product feature though.

File: xonsh/parsers/base.py (hunk 1 above)
Comment (scopatz) [reply xonsh/parsers/base.py]: I think that isn't possible given how the parser reads tokens from left to right, but it is good to be robust here.  Fix incoming.

Comment (scopatz) [reply xonsh/parsers/base.py]: Fix pushed up!

**Comment by melund** [comment]
I won't say I understand the PLY stuff. But I believe @gforsyth looked it that part through.

**Pull request merged**